anndata/core/anndata.py

```diff
@@ -2194,11 +2194,34 @@ def concatenate(
                     sparse_format_l = sparse_layers[0].getformat()
                     layers[key] = layers[key].asformat(sparse_format_l)
 
+        # raw
+        num_raws = sum(1 for _adata in all_adatas if _adata.raw is not None)
+        if num_raws == 0:
+            new_adata_raw = None
+        elif num_raws == len(all_adatas):
+            _raws = [_adata.raw.to_adata() for _adata in all_adatas]
+            new_adata_raw = self.concatenate(
+                *_raws,
+                join=join,
+                batch_key=batch_key,
+                batch_categories=batch_categories,
+                index_unique=index_unique,
+            )
+        else:
+            logger.warning(
+                'Only some adata objects have `.raw` attribute, not concatenating them.'
+            )
+            new_adata_raw = None
```

Comment:
Instead of this, could the anndata be constructed first, then add the raw in the conditional? E.g.:

```python
have_raw = [getattr(a, "raw", None) is not None for a in all_adatas]
if all(have_raw):
    ...
    adata.raw = new_adata_raw
elif any(have_raw):
    # Just throw warning
```